include/kernel.h

```diff
@@ -0,0 +1,534 @@
+ * killed.
+ *
+ * @param[in] name
+ *            Name of the thread (for debugging purposes)
+ * @param[in] stack_size
+ *            Size of the stack in bytes. Minimum suggested size
+ *            is 2048.
+ * @param[in] pri
+ *            Priority of the thread (0-127). Negative values are not allowed.
+ *            Higher number means higher priority.
+ * @param[in] user_entry
+ *            Entry point of the thread. This is the function that will
+ *            be called when the thread begins execution. If this function
+ *            ever returns, the thread is automatically killed.
+ * @param[in] user_data
+ *            Argument that will be passed to the entry point of the thread.
+ *
+ * @return    A pointer to the new thread. It is not necessary to store
+ *            this reference if not required; the thread will clean up
+ *            after itself when it exits.
+ */
+kthread_t* kthread_new(const char *name, int stack_size, int8_t pri,
+	void (*user_entry)(void*), void *user_data);
+
+
+/**
+ * @brief Yield execution of the current thread and run the scheduler.
+ *
+ * This function allows the current thread to cooperatively yield
+ * its execution to allow other threads to run.
+ *
+ * The scheduler will switch to the highest priority thread that is
+ * currently ready to run. If no ready thread has a priority higher than
+ * the current thread, the scheduler will switch to a different thread 
+ * of the same priority of the current one (the scheduler will guarantee
+ * a correct round-robin scheduling among threads of the same priority).
+ * If no ready thread has a priority higher or equal than the current
+ * one, the scheduler will reschedule the current thread, that will
+ * continue execution.
+ *
+ * @note The scheduler is semi-preemptive. Any interrupt could cause
+ * a thread switch to happen if the interrupt itself makes a thread
+ * become ready. If you need a block of code to be executed without
+ * any context switch, make sure to disable interrupts.
+ */
+void kthread_yield(void);
+
+
+/**
+ * @brief Sleep for the specified interval, allowing execution of other threads.
+ *
+ * This function will put the current thread to sleep for a specified
+ * time interval, allowing other threads to run.
+ *
+ * The sleeping interval is expressed in hardware ticks. See #TICKS_READ
+ * for an explanation of hardware ticks, and #TICKS_PER_SECOND and #TICKS_FROM_MS
+ * for constants and macros that simplify conversion to standard time units.
+ *
+ * Notice that #wait_ticks and #wait_ms are not threading aware, so they will
+ * just spin-wait while keeping the current thread ready. Other higher-priority
+ * threads might still be scheduled if they are awaken by an interrupt, but in
+ * general thread_sleep is a better primitive if you want the current thread
+ * to yield.
+ *
+ * @note This function requires the timer module, so #timer_init must have
+ *       been called.
+ *
+ * @param[in] ticks
```

Comment:
This won't guarantee a re-scheduling after the time is elapsed though right? o/w there should be another rule for switching threads, other than priority.
We would probably need to change priority if we want that behavior.

Reply:
> This won't guarantee a re-scheduling after the time is elapsed though right? 

It guarantees that the thread becomes ready again after the time is elapsed. Whether it is scheduled or not at that point depends on what are threads are ready. Consider also that even if it's scheduled again at that point, it might be switched off one microsecond later once a higher priority thread becomes ready. So in general `sleep` in a threading context means `sleep at least`, given that any thread can be interrupted by any other higher priority thread at any point.

Reply:
The rules above makes is really clear now.